cpp/src/arrow/util/iterator.cc

```diff
@@ -0,0 +1,175 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/iterator.h"
+
+#include <condition_variable>
+#include <cstdint>
+#include <deque>
+#include <mutex>
+#include <thread>
+
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace detail {
+
+ReadaheadPromise::~ReadaheadPromise() {}
+
+class ReadaheadQueue::Impl : public std::enable_shared_from_this<ReadaheadQueue::Impl> {
+ public:
+  explicit Impl(int64_t readahead_queue_size) : max_readahead_(readahead_queue_size) {}
+
+  ~Impl() { EnsureShutdownOrDie(false); }
+
+  void Start() {
+    // Cannot do this in constructor as shared_from_this() would throw
+    DCHECK(!thread_.joinable());
+    auto self = shared_from_this();
+    thread_ = std::thread([self]() { self->DoWork(); });
+    DCHECK(thread_.joinable());
+  }
+
+  void EnsureShutdownOrDie(bool wait = true) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (!please_shutdown_) {
+      ARROW_CHECK_OK(ShutdownUnlocked(std::move(lock), wait));
+    }
+    DCHECK(!thread_.joinable());
+  }
+
+  Status Append(std::unique_ptr<ReadaheadPromise> promise) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (please_shutdown_) {
+      return Status::Invalid("Shutdown requested");
+    }
+    todo_.push_back(std::move(promise));
+    if (static_cast<int64_t>(todo_.size()) == 1) {
```

Comment:
I assume if there were tasks in todo_ already then the worker will be woken up by design